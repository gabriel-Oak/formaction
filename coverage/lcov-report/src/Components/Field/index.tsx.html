<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/Components/Field/index.tsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="..\..\..\prettify.css" />
    <link rel="stylesheet" href="..\..\..\base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(..\..\..\sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="..\..\..\index.html">All files</a> / <a href="index.html">src/Components/Field</a> index.tsx
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>59/59</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>14/14</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>11/11</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>44/44</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130</td><td class="line-coverage quiet"><span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">14x</span>
<span class="cline-any cline-yes">14x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import * as React from 'react';
import { FormContext } from '../../shared/context';
import FieldWapper from './FieldWrapper';
&nbsp;
export interface InputEvent {
  target: {
    value: string;
  };
  type: string;
  nativeEvent: any;
  preventDefault: Function;
}
&nbsp;
export type FieldProps = {
  name: string;
  renderComponent?: React.FC | any;
  validators?: Array&lt;(value: any) =&gt; string | undefined&gt;;
}
&nbsp;
class Field extends React.PureComponent&lt;FieldProps&gt; {
&nbsp;
  static contextType = FormContext;
&nbsp;
  constructor(props: FieldProps) {
    super(props);
&nbsp;
    this.onChange = this.onChange.bind(this);
    this.handleNativeChange = this.handleNativeChange.bind(this);
    this.handleNativeBlur = this.handleNativeBlur.bind(this);
  }
&nbsp;
  componentDidMount() {
    const { fields, form } = this.context;
    const { name } = this.props;
&nbsp;
    fields[name] = {
      touched: false,
      errors: this._getErrors(form[name])
    }
  }
&nbsp;
  onChange(value: any) {
    const { updateForm } = this.context;
&nbsp;
    updateForm(this._formatForm(value));
  }
&nbsp;
  handleNativeChange(event: InputEvent) {
    const { updateForm } = this.context;
    const { target: { value } } = event;
&nbsp;
    updateForm(this._formatForm(value));
  }
&nbsp;
  handleNativeBlur(_event: InputEvent) {
    const { fields, setFields } = this.context;
    const { name } = this.props;
&nbsp;
    setFields({
      ...fields,
      [name]: {
        ...fields[name],
        touched: true
      }
    });
  }
&nbsp;
  _formatForm(value: any): any {
    const { fields, setFields, form } = this.context;
    const { name } = this.props;
&nbsp;
    const tempForm = JSON.parse(JSON.stringify(form));
&nbsp;
    setFields({
      ...fields,
      [name]: {
        ...fields[name],
        errors: this._getErrors(value)
      }
    });
&nbsp;
    if (value) {
      tempForm[name] = value;
    } else {
      delete tempForm[name];
    }
&nbsp;
    return tempForm;
  }
&nbsp;
  _getErrors(value: any): any[] {
    let validation = this.props.validators || [];
    return validation
      .map(validate =&gt; validate(value))
      .filter(error =&gt; error);
  }
&nbsp;
  render(): React.ReactNode {
    const { name, renderComponent, validators, children, ...rest } = this.props;
    const { form, fields } = this.context;
&nbsp;
    return (
      &lt;&gt; {
        renderComponent
          ?
          FieldWapper(renderComponent, {
            onChange: this.onChange,
            meta: fields[name] || {},
            input: {
              ...rest,
              name,
              value: form[name] || '',
              onChange: this.handleNativeChange,
              onBlur: this.handleNativeBlur
            }
          })
          :
          &lt;input
            {...rest}
            name={name}
            value={form[name] || ''}
            onChange={this.handleNativeChange}
          /&gt;
      } &lt;/&gt;
    );
  }
}
&nbsp;
export default Field
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Dec 19 2019 18:17:07 GMT-0200 (GMT-02:00)
</div>
</div>
<script src="..\..\..\prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="..\..\..\sorter.js"></script>
</body>
</html>
