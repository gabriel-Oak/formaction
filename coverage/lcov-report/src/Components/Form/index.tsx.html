<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/Components/Form/index.tsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="..\..\..\prettify.css" />
    <link rel="stylesheet" href="..\..\..\base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(..\..\..\sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="..\..\..\index.html">All files</a> / <a href="index.html">src/Components/Form</a> index.tsx
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>31/31</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>7/7</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>5/5</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>31/31</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">/**
 * @component Form
 */
import * as React from 'react'
import FormHooks, { FormProps } from './hooks';
import { FormContext } from '../../shared/context';
&nbsp;
&nbsp;
const Form: React.FC&lt;FormProps&gt; = (props: FormProps) =&gt; {
  const {
    form,
    updateForm,
    useEffect,
    fields,
    setFields,
    children,
    onSubmit,
    onChange,
    values,
    ...rest
  } = FormHooks(props);
&nbsp;
&nbsp;
  useEffect(() =&gt; {
    !values &amp;&amp; onChange &amp;&amp; onChange(form);
  }, [form]);
&nbsp;
  useEffect(() =&gt; {
    setFields(fields);
  }, [fields]);
&nbsp;
  const handleSubmit = (event: React.FormEvent&lt;HTMLFormElement&gt;) =&gt; {
    event.preventDefault();
    const tempFields = JSON.parse(JSON.stringify(fields));
&nbsp;
    const inValid = Object
      .keys(tempFields)
      .some(
        key =&gt; {
          tempFields[key].touched = true;
          return tempFields[key].errors.length;
        }
      );
&nbsp;
    if (inValid) {
      setFields(tempFields);
      return;
    }
&nbsp;
    onSubmit(form);
  }
&nbsp;
  return (
    &lt;form
      {...rest}
      onSubmit={handleSubmit}
    &gt;
      &lt;FormContext.Provider value={{
        form: values || form,
        updateForm,
        fields,
        setFields
      }}&gt;
&nbsp;
        {children}
&nbsp;
      &lt;/FormContext.Provider&gt;
    &lt;/form&gt;
  );
}
&nbsp;
export default Form;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Dec 19 2019 16:35:04 GMT-0200 (GMT-02:00)
</div>
</div>
<script src="..\..\..\prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="..\..\..\sorter.js"></script>
</body>
</html>
