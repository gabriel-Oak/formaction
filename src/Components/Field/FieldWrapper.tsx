import * as React from 'react';

interface FieldProps {
  name: string;
  onChange: Function;
  value?: string;
  className?: string;
  disabled?: boolean;
  readonly?: boolean;
}

const FieldWapper = (RenderField: any, props: FieldProps) => (
  <RenderField {...props} />
);

export default FieldWapper;
