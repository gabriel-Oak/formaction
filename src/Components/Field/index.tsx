import * as React from 'react';
import { FormContext } from '../../shared/context';
import FieldWapper from './FieldWrapper';

export interface InputEvent {
  target: {
    value: string;
  };
  type: string;
  nativeEvent: any;
  preventDefault: Function;
}

export type FieldProps = {
  name: string;
  renderComponent: React.FC | any;
  validators?: Array<(value: any) => string | undefined>;
}

class Field extends React.PureComponent<FieldProps> {

  static contextType = FormContext;

  constructor(props: FieldProps) {
    super(props);

    this.onChange = this.onChange.bind(this);
    this.handleNativeField = this.handleNativeField.bind(this);
  }

  componentDidMount() {
    const { fields, form } = this.context;
    const { name, validators } = this.props;
    const validation = validators || [];

    fields[name] = {
      touched: false,
      errors: validation
        .map(validate => validate(form[name]))
        .filter(error => error)
    }
  }

  onChange(value: any) {
    const { setForm } = this.context;

    setForm(this._formatForm(value));
  }

  handleNativeField(event: InputEvent) {
    const { setForm } = this.context;
    const { target: { value } } = event;

    setForm(this._formatForm(value));
  }

  _formatForm(value: any): any {
    const { fields, setFields, form } = this.context;
    const { name, validators } = this.props;
    const validation = validators || [];

    const tempForm = JSON.parse(JSON.stringify(form));

    setFields({
      ...fields,
      [name]: {
        touched: true,
        errors: validation
          .map(validate => validate(value))
          .filter(error => error)
      }
    });

    if (value) {
      tempForm[name] = value;
    } else {
      delete tempForm[name];
    }

    return tempForm;
  }

  render(): React.ReactNode {
    const { name, renderComponent, validators, children, ...rest } = this.props;
    const { form, fields } = this.context;

    return (
      <> {
        renderComponent
          ?
          FieldWapper(renderComponent, {
            onChange: this.onChange,
            meta: fields[name],
            input: {
              ...rest,
              name,
              value: form[name] || '',
              onChange: this.handleNativeField
            }
          })
          :
          <input
            {...rest}
            name={name}
            value={form[name] || ''}
            onChange={this.handleNativeField}
          />
      } </>
    );
  }
}

export default Field
