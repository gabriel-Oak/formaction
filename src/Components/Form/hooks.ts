import { useState, useEffect, ReactNode } from "react";

export type FormProps = {
  children: ReactNode;
  initialValues?: any;
  onSubmit: (values: any) => {};
  onChange?: (values: any) => {};
}


const FormHooks = (props: FormProps) => {
  const { initialValues, ...rest } = props;
  const [form, setForm] = useState(initialValues || {});
  const [fields, setFields] = useState({});


  return {
    form,
    setForm,
    fields,
    setFields,
    useEffect,
    ...rest
  }
};

export default FormHooks;
