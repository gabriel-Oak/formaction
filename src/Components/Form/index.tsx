/**
 * @component Form
 */
import * as React from 'react'
import FormHooks, { FormProps } from './hooks';
import { FormContext } from '../../shared/context';


const Form: React.FC = (props: FormProps) => {
  const {
    form,
    setForm,
    useEffect,
    fields,
    setFields,
    children,
    onSubmit,
    onChange,
    ...rest
  } = FormHooks(props);


  useEffect(() => {
    onChange && onChange(form);
  }, [form]);

  useEffect(() => {
    setFields(fields);
  }, [fields]);

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const tempFields = JSON.parse(JSON.stringify(fields));

    const inValid = Object
      .keys(tempFields)
      .some(
        key => {
          tempFields[key].touched = true;
          return tempFields[key].errors.length;
        }
      );

    if (inValid) {
      setFields(tempFields);
      return;
    }

    onSubmit(form);
  }


  return (
    <form
      {...rest}
      onSubmit={handleSubmit}
    >
      <FormContext.Provider value={{
        form,
        setForm,
        fields,
        setFields
      }}>

        {children}

      </FormContext.Provider>
    </form>
  );
}

export default Form;
