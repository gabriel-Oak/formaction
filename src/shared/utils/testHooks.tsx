import * as React from 'react';
import * as renderer from 'react-test-renderer';


export default function testHook(runHook: any) {
  const HookWrapper: React.FC = () => {
    return <span {...runHook()} />;
  }

  const wrapper = renderer
    .create(<HookWrapper />)
    .toJSON();

  return wrapper && wrapper.props;
}
