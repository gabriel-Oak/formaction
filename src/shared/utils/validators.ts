export const isEmail = (value: string) =>
  !value || !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value) ?
    'Esse campo deve ser do tipo e-mail' : undefined;

export const isNumber = (value: any) => (
  !value || !isNaN(+value)
    ? undefined
    : 'Esse campo deve ser do tipo número'
);

export const isRequired = (value?: any) => (
  value || typeof value === 'number'
    ? undefined
    : 'Esse campo é obrigatorio'
);
