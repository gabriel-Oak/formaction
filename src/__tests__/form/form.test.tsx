import * as React from 'react';
import FormHooks from '../../Components/Form/hooks';
import testHook from '../../shared/utils/testHooks';
import { act, create, ReactTestRenderer, ReactTestRendererTree } from 'react-test-renderer';
import Form from '../../Components/Form';
import Field from '../../Components/Field';
import { isEmail } from '../../shared/utils/validators';

describe('Form', () => {
  let wrapper: ReactTestRenderer;
  let onSubmit: () => {};
  const onChange = jest.fn();
  const formMock = { honney: 'pot' };

  const getSubmit = (tree: ReactTestRendererTree | null) => {
    const form = tree && tree.rendered;
    const props = form && form.props;
    return props && props.onSubmit
  }

  beforeEach(() => {
    onSubmit = jest.fn();
    wrapper = create(
      <Form
        onSubmit={onSubmit}
        onChange={onChange}
      >
        <Field
          name='testField'
          label='Testing is fun'
        />
      </Form>
    );
  });

  it('should match snapshot', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should submit form', () => {
    const submit = getSubmit(wrapper.toTree());

    act(() => {
      submit({ preventDefault: jest.fn() });
    });

    expect(onSubmit).toHaveBeenCalled();
  });

  it('should not submit form', () => {
    wrapper = create(
      <Form
        onSubmit={onSubmit}
        onChange={onChange}
      >
        <Field name='testField' validators={[isEmail]} />
      </Form>
    );

    const submit = getSubmit(wrapper.toTree());

    act(() => {
      submit({ preventDefault: jest.fn() });
    });

    expect(onSubmit).toHaveBeenCalledTimes(0);
  });

  describe('Hooks', () => {
    let hooks: any;

    beforeEach(() => {
      hooks = testHook(
        () => FormHooks({
          onSubmit: jest.fn(),
          children: []
        })
      );
    })

    it('should init the form with default values', () => {
      expect(hooks.form).toEqual({});
    });

    it('should init the form with pre values', () => {
      hooks = testHook(
        () => FormHooks({
          onSubmit: jest.fn(),
          children: [],
          initialValues: {
            test: 'Jest'
          }
        })
      );

      expect(hooks.form).toEqual({ test: 'Jest' });
    });

    it('should update form values', () => {
      act(() => {
        hooks.updateForm(formMock);
      });
    });

    it('should update controlled form values', () => {
      const onChange = jest.fn();

      hooks = testHook(
        () => FormHooks({
          onSubmit: jest.fn(),
          onChange,
          children: [],
          values: {
            test: 'Jest'
          }
        })
      );

      act(() => {
        hooks.updateForm(formMock);
      });

      expect(onChange).toHaveBeenCalledWith(formMock);
    });

  });
});
