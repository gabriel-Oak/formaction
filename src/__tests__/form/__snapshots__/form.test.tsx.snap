// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match snapshot 1`] = `
<form
  onSubmit={[Function]}
>
   
  <label
    htmlFor="testField"
  >
    Testing is fun
  </label>
  <input
    name="testField"
    onChange={[Function]}
    value=""
  />
   
</form>
`;
